Lib/collections/__init__.py

```diff
@@ -412,24 +434,69 @@ def namedtuple(typename, field_names, *, verbose=False, rename=False, module=Non
             raise ValueError('Encountered duplicate field name: %r' % name)
         seen.add(name)
 
-    # Fill-in the class template
-    class_definition = _class_template.format(
-        typename = typename,
-        field_names = tuple(field_names),
-        num_fields = len(field_names),
-        arg_list = repr(tuple(field_names)).replace("'", "")[1:-1],
-        repr_fmt = ', '.join(_repr_template.format(name=name)
-                             for name in field_names),
-        field_defs = '\n'.join(_field_template.format(index=index, name=name)
-                               for index, name in enumerate(field_names))
-    )
-
-    # Execute the template string in a temporary namespace and support
-    # tracing utilities by setting a value for frame.f_globals['__name__']
-    namespace = dict(__name__='namedtuple_%s' % typename)
-    exec(class_definition, namespace)
-    result = namespace[typename]
-    result._source = class_definition
+    arg_list = repr(tuple(field_names)).replace("'", "")[1:-1]
+    num_fields = len(field_names)
+    repr_fmt = '(' + ', '.join(_repr_template.format(name=name)
+                               for name in field_names) + ')'
+
+    namespace = {'_tuple': tuple}
+    new_source = _new_template.format(typename=typename, arg_list=arg_list)
+    exec(new_source, namespace)
+    __new__ = namespace['__new__']
+
+    @classmethod
+    def _make(cls, iterable, new=tuple.__new__, len=len):
+        result = new(cls, iterable)
+        if len(result) != cls._num_fields:
```

Comment:
I didn't want to introduce another function call. That might be premature optimization though.